File: Easy/Product array puzzle/product-array-puzzle.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*; 

class GFG{
    public static void main(String args[]) throws IOException { 
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while(t > 0){
        	int n = sc.nextInt();
        	int[] array = new int[n];
        	for (int i=0; i<n ; i++ ) {
        		array[i] = sc.nextInt();
        	}
            Solution ob = new Solution();
            long[] ans = new long[n];
            ans = ob.productExceptSelf(array, n); 

           	for (int i = 0; i < n; i++) { 
				System.out.print(ans[i]+" ");
			} 
            System.out.println();
            t--;
        }
    } 
} 
  


// } Driver Code Ends


//User function Template for Java


class Solution 
{ 
public static long[] productExceptSelf(int nums[], int n) 
{ 
        // code here
        long[] arr = new long[n];
        int count = 0;
        long prod = 1;
        int j=-1;
        for(int i=0; i< n;i++){
            if(nums[i] != 0) {
                prod = prod*nums[i];
            }  
                
            else{
                j=i;
                count++;
            }
                
        }
        
        if(count == 0){
            for(int i=0;i<n;i++){
                arr[i] = prod/nums[i];
            }
        }
        else if(count == 1){
            arr[j] = prod;
        }
        
        return arr;
} 
} 

File: Difficulty: Easy/Product array puzzle/product-array-puzzle.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++


class Solution {
  public:
    vector<long long int> productExceptSelf(vector<long long int>& nums) {

        vector<long long int>ans;
        long long int zeroCount=0;
        long long int mult=1;
        for(int i=0;i<nums.size();i++)
        {
            if(nums[i]!=0)
            mult*=nums[i];
            else
            zeroCount++;
        }
        for(int i=0;i<nums.size();i++){
            if(nums[i]==0){
                if(zeroCount==1)
                ans.push_back(mult);
                else
                    ans.push_back(0);
            }
            else{
                if(zeroCount)
                ans.push_back(0);
                else
                ans.push_back(mult/nums[i]);
            }
            
        }
        return ans;
    }
};




//{ Driver Code Starts.
int main() {
    int t; // number of test cases
    cin >> t;
    while (t--) {
        int n; // size of the array
        cin >> n;
        vector<long long int> arr(n), vec(n);

        for (int i = 0; i < n; i++) // input the array
        {
            cin >> arr[i];
        }
        Solution obj;
        vec = obj.productExceptSelf(arr); // function call

        for (int i = 0; i < n; i++) // print the output
        {
            cout << vec[i] << " ";
        }
        cout << endl;
    }
    return 0;
}
// } Driver Code Ends